src/sipmpdf/functions.py

```diff
@@ -216,47 +245,384 @@ def darkcurrent_response(x: np.float64,
+                       darkcurrent_response_original(x, gain, resolution), 0.0)
+
+  outer = normal(x, 0, coarse) + normal(x, gain, coarse)
+  outer_norm = (kern.log(coarse_res) - kern.log(resolution)) / 2.0 / (
+    kern.log(1 - resolution) - kern.log(resolution))
+  return nominal + outer * outer_norm
 
 
 def ap_response_smeared(x: np.float64, smear: np.float64, n_ap: np.int64,
                         beta: np.float64) -> np.float64:
-  kern = kernel_switch(x, smear, n_np, beta)
-  pass
+  """
+  Afterpulse response given a fixed number of after pulses occurs accounting for the effects of random noise.
+  
+  Parameters
+  ----------
+  x : np.float64
+      Observation
+  smear: np.float64
+      Value by which the response is smeared by
+  n_ap : np.float64
+      Number of afterpulses that occur
+  beta : np.float64
+      Afterpulse timescale factor.
+  Returns
+  -------
+  np.float64
+      Probability density of the after pulse response
+  """
+  kern = kernel_switch(x, smear, n_ap, beta)
+  return kern.where(
+    n_ap <= 1.5,
+    kern.exp(-x / beta) *
+    (1 + kern.erf(x / (smear * kern.sqrt(kern.tofloat64(2))))) / (2 * beta),
+    ap_response(x=x, n_ap=n_ap, beta=beta))
 
 
+@expand_shape
 def darkcurrent_response_smeared(x: np.float64,
                                  smear: np.float64,
-                                 pedestal: np.float64,
                                  gain: np.float64,
                                  resolution: np.float64 = 1e-4) -> np.float64:
-  kern = kernel_switch(x, smaer, pedestal, gain, resolution)
-  pass
-
-
-def sipm_response(
-  x: np.float64,
-  pedestal: np.float64,
-  gain: np.float64,
-  common_noise: np.float64,
-  pixel_noise: np.float64,
-  poisson_mean: np.float64,
-  poisson_borel: np.float64,
-  ap_prob: np.float64,
-  ap_beta: np.float64,
-) -> np.float64:
+  """
+  Dark current response with random noise, given a detector resolution factor
+  (calculated relative to the gain of the system)
+
+  Parameters
+  ----------
+  x : np.float64
+      Observation
+  smear : np.float64
+      The value by which the dark current response is smeared
+  gain : np.float64
+      The gain of the SiPM
+  resolution : np.float64, optional
+      resolution factor, relative to the SiPM gain factor, by default 1e-4
+
+  Returns
+  -------
+  np.float64
+      Probability density of observing a dark current response
+  """
+
+  kern = kernel_switch(x, smear, gain, resolution)
+
+  n_samples = 1024
+
+  #extend arrays
+  def extend_arr(x):
+    return kern.repeat(x[kern.newaxis, ...], n_samples, axis=0)
+
+  x = extend_arr(x)
+  smear = extend_arr(smear)
+  gain = extend_arr(gain)
+  resolution = extend_arr(resolution)
+
+  delta = 8 * smear / n_samples  # Getting the integration window
+  idx = kern.local_index(x, axis=0)
+  x_pr = -4 * smear + delta * (idx + 0.5)
+
+  # Given that the computed result will exhibit periodicity within the
```

Comment:
Remove this comment, as we this comment is not descriptive of what we are doing.